jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/ProxyProvider.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.maven;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.maven.settings.Proxy;
+import org.apache.maven.settings.Settings;
+
+/** Initializes and retrieves proxy settings from either Maven settings or system properties * */
+public class ProxyProvider {
+
+  /**
+   * Initializes proxy settings based on Maven settings and system properties.
+   *
+   * @param settings - Maven settings from mojo
+   */
+  public static void init(Settings settings) {
+    List<Proxy> activeProxies =
+        settings
+            .getProxies()
+            .stream()
+            .filter(proxy -> proxy.isActive())
+            .collect(Collectors.toList());
+    for (Proxy proxy : activeProxies) {
+      propagateProxyProperties(proxy);
+    }
+  }
+
+  /**
+   * Propagate Maven proxy properties into system properties to be picked up by Connections.
+   *
+   * @param proxy Maven proxy
+   */
+  private static void propagateProxyProperties(Proxy proxy) {
+    if (proxy.getProtocol().equalsIgnoreCase("http")) {
+      propagateProxyProperty("http.proxyHost", proxy.getHost());
+      propagateProxyProperty("http.proxyPort", String.valueOf(proxy.getPort()));
+      propagateProxyProperty("http.proxyUser", proxy.getUsername());
+      propagateProxyProperty("http.proxyPassword", proxy.getPassword());
+    } else if (proxy.getProtocol().equalsIgnoreCase("https")) {
+      propagateProxyProperty("https.proxyHost", proxy.getHost());
+      propagateProxyProperty("https.proxyPort", String.valueOf(proxy.getPort()));
+      propagateProxyProperty("https.proxyUser", proxy.getUsername());
+      propagateProxyProperty("https.proxyPassword", proxy.getPassword());
+    }
+    propagateProxyProperty("http.nonProxyHosts", proxy.getNonProxyHosts());
+  }
+
+  /**
+   * Propagate Maven proxy property into system property to be picked up by Connections. Only set
+   * the system property if not already set.
+   *
+   * @param name proxy system property name
+   * @param value proxy system property value
+   */
+  private static void propagateProxyProperty(String name, String value) {
+    if (value != null && System.getProperty(name) == null) {
```

Comment:
I think the proxy updating should happen all or not. For example, someone can define `https.proxyHost` from the command line but leave `https.proxyPort` undefined, in which case the default port will be chosen. It is also possible to leave `https.proxyUser` if the proxy doesn't require auth. So, I think the logic should basically be

```
if (only when none of http.x properties are defined) {
  Proxy firstActiveHttpProxy = <first active HTTP-targeting proxy from settings.xml>;
  System.setProperty(http.proxyHost, firstActiveHttpProxy.getHost());
  System.setProperty(http.proxyPort, firstActiveHttpProxy.getPort());
  ...
}
if (only when none of https.x properties are defined) {
  Proxy firstActiveHttpsProxy = <first active HTTPS-targeting proxy from settings.xml>;
  System.setProperty(https.proxyHost, firstActiveHttpsProxy.getHost());
  System.setProperty(https.proxyPort, firstActiveHttpsProxy.getPort());
  ...
}
```

Reply:
Ok, I will refactor with approach of first protocol maven proxy wins, unless one of protocol system properties were already set.
FYI, I am thinking of moving to custom proxy selector approach where eligible maven proxies are added to proxy map inside custom RegistryProxySelector. This RegistryProxySelector can then be used by Connection / HttpTransport, to select proxy matching registry url, and register proxy credentials on transport where needed.

Reply:
> Ok, I will refactor with approach of first protocol maven proxy wins, unless one of protocol system properties were already set.

Sound good.

> This RegistryProxySelector can then be used by Connection / HttpTransport, to select proxy matching registry url

I honestly think you don't have to do this. I don't see you can configure per-host (or per-container-registry) proxies through `settings.xml`. You change the current a little bit, and I think we're good to go.

Reply:
No it would still be max 1 proxy per protocol. But I thought it could be a bit more elegant way to transfer proxy config from maven plugin into core. If you tell me not to bother I will only do the small refactor. In any case I probably will not be able to finish/test before Monday.

Reply:
Yeah, don't be bothered. In this case, I think we can keep things simple.

Reply:
Ok I will commit today but test finally on Monday